### PR TITLE
Migrate to pyproject.toml + cleanup

### PR DESCRIPTION
**Pull request by Saransh-cpp**
- Zarr has a lot of redundant files in the root directory. The information contained in these files can be easily moved to `pyproject.toml`.
- Latest Python PEPs encourage users to get rid of `setup.py` and switch to `pyproject.toml`
- We should not be using `setup.cfg` until it is a necessity - https://github.com/pypa/setuptools/issues/3214
- We should not be using `setuptools` as a frontend (`python setup.py install`) - this is not maintained (confirmed by `setuptools` developers, but I cannot find the exact issue number at this moment)
- Zarr should switch to pre-commit.ci and remove the `pre-commit` workflow

I have tried to perform a 1:1 port. No extra information was added and all the existing metadata has been moved to `pyproject.toml`.

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by Saransh-cpp** [comment]
A fix for the failing docs - https://github.com/readthedocs/readthedocs.org/issues/3322#issuecomment-347188196

**Comment by Saransh-cpp** [comment]
Ideally, the `requirements*.txt` files should also be added as optional dependencies in `pyproject.toml`, but that would make this PR even bigger. Maybe I can carry that out in a follow-up PR. (That would also require editing documentation.)

**Comment by joshmoore** [comment]
Thanks for starting this discussion, @Saransh-cpp.

Just a heads up that the readthedocs failure will need to be fixed on their side so isn't something to worry about:

```
/home/docs/checkouts/readthedocs.org/user_builds/zarr/envs/1158/bin/python: can't open file '/home/docs/checkouts/readthedocs.org/user_builds/zarr/checkouts/1158/setup.py': [Errno 2] No such file or directory
```

File: docs/release.rst [hunk 1]
```diff
@@ -23,6 +23,12 @@ Release notes
   data.
   By :user:`Richard Shaw <jrs65>` :issue:`843`.
 
+Maintenance
```
Comment (joshmoore) [inline docs/release.rst:26]: 2.13.3 went out. Can you move this to unreleased?

**Review by joshmoore — CHANGES_REQUESTED** [review]
Thanks for starting this discussion, @Saransh-cpp. I'm :+1: for use of pre-commit.ci and the readthedocs failure will need fixing in the RTD settings so the single test failure can be safely ignored for the moment.

> I have tried to perform a 1:1 port. No extra information was added and all the existing metadata has been moved to pyproject.toml.

Thanks, this is a great way to keep the PR reviewable. There's one minor change from my side, but otherwise generally approve of moderization. :smile:

Do any other @zarr-developers/python-core-devs have opinions here?

**Comment by Saransh-cpp** [comment]
Thanks for the quick review, @joshmoore! Should I go ahead and remove the `Pre-commit-hooks` workflow from zarr?

**Comment by joshmoore** [comment]
> Should I go ahead and remove the `Pre-commit-hooks` workflow from zarr?

Yes, please, so that anyone who is reviewing will also be aware of the change.

**Comment by Saransh-cpp** [comment]
TODO for maintainers -
- [ ] Turn on pre-commit.ci for zarr

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,77 @@
...
+    'Programming Language :: Python :: 3.8',
+    'Programming Language :: Python :: 3.9',
+    'Programming Language :: Python :: 3.10',
+]
+
+[project.optional-dependencies]
+jupyter = [
+    'notebook',
+    'ipytree',
+]
```
Comment (jakirkham) [inline pyproject.toml:40]: Heads up. This just got [updated]( https://github.com/zarr-developers/zarr-python/blob/f9581cb6db4c2950e200add3785bac2685aac391/setup.py#L31-L35 )

```suggestion
[project.optional-dependencies]
jupyter = [
    'notebook',
    'ipytree>=0.2.2',
    'ipywidgets>=8.0.0',
]
```

File: pyproject.toml [hunk 3]
```diff
@@ -1,3 +1,77 @@
 [build-system]
-requires = ["setuptools>=40.8.0", "wheel"]
+requires = ["setuptools", "setuptools-scm", "wheel"]
```
Comment (jakirkham) [inline pyproject.toml:2]: Should we still have the lower bound on `setuptools`?

File: pyproject.toml [hunk 4]
```diff
@@ -1,3 +1,77 @@
...
 build-backend = "setuptools.build_meta"
+
+
+[project]
+name = "zarr"
+description = "An implementation of chunked, compressed, N-dimensional arrays for Python"
+readme = { file = "README.md", content-type = "text/markdown" }
+maintainers = [ {name = "Alistair Miles", email = "alimanfoo@googlemail.com"} ]
```
Comment (jakirkham) [inline pyproject.toml:10]: Would it make sense to break these over multiple lines as done below with other sections?

**Comment by Saransh-cpp** [comment]
Thanks for the quick review, @jakirkham!

> Would it make sense to break these over multiple lines as done below with other sections?

Unfortunately, [tomli](https://pypi.org/project/tomli/) cannot parse blocks like -
```
readme = {
    file = "README.md", content-type = "text/markdown"
}
```
Hence, anything the metadata written in between `{ ... }` should be written in a single line.

File: pyproject.toml [hunk 5]
```diff
@@ -1,3 +1,80 @@
 [build-system]
-requires = ["setuptools>=40.8.0", "wheel"]
+requires = ["setuptools>=40.8.0", "setuptools-scm", "wheel"]
 build-backend = "setuptools.build_meta"
+
+
```
Comment (jakirkham) [inline pyproject.toml:5]: Maybe this should be a single blank line? Alternatively suppose we could use two blank lines between all sections

```suggestion

```

File: release.txt [hunk 6]
```diff
@@ -3,5 +3,6 @@ tox
 echo $version
 git tag -a v$version -m v$version
 git push --tags
-python setup.py register sdist
+python -m pip install build
+python -m build
```
Comment (jakirkham) [inline release.txt:7]: Would suggest not installing things like this in the users environment and instead document how to do it either here or elsewhere.

```suggestion
# Install `build` if not present with `python -m pip install build` or similar
python -m build
```

File: release.txt [hunk 7]
```diff
@@ -3,5 +3,6 @@ tox
...
 twine upload dist/zarr-${version}.tar.gz
```
Comment (jakirkham) [inline release.txt:8]: We may want to include `twine` in that documentation

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -1,3 +1,6 @@
+ci:
+  autoupdate_commit_msg: "chore: update pre-commit hooks"
+  autofix_commit_msg: "style: pre-commit fixes"
```
Comment (jakirkham) [inline .pre-commit-config.yaml:3]: Will these perform the style fixes for us after pushing a PR? If so, that's great 😄

Though we probably want to document this behavior somewhere so users know about this. Maybe in [the contributor guide]( https://zarr.readthedocs.io/en/stable/contributing.html )?

Comment (Saransh-cpp) [reply .pre-commit-config.yaml:3]: Yes! But the maintainers should install the `precommit.ci` app (https://github.com/marketplace/pre-commit-ci) for Zarr's GH repository.

I'll document it in the contributor guide.

**Comment by Saransh-cpp** [comment]
Can I somehow turn the lgtm bot off for this PR? I feel like it has been spamming more than helping 😅

**Comment by Saransh-cpp** [comment]
I have added docs for `pre-commit`! The last remaining thing is activating the pre-commit bot for `Zarr` - https://github.com/marketplace/pre-commit-ci

**Comment by joshmoore** [comment]
> Can I somehow turn the lgtm bot off for this PR? I feel like it has been spamming more than helping 😅

You're definitely right, though I do wonder why it's shown up here. Perhaps @DimitriPapadopoulos would know. I'll delete some of the superfluous ones and we'll go from there.

**Review by joshmoore — APPROVED** [review]

File: docs/contributing.rst [hunk 9]
```diff
@@ -179,14 +179,33 @@ also collected automatically via the Codecov service, and total
...
-~~~~~~~~~~~~~~
+Code standards - using pre-commit
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 All code must conform to the PEP8 standard. Regarding line length, lines up to 100
 characters are allowed, although please try to keep under 90 wherever possible.
-Conformance can be checked by running::
 
-    $ python -m flake8 --max-line-length=100 zarr
+``Zarr`` uses a set of ``pre-commit`` hooks and the ``pre-commit`` bot to format,
```
Comment (joshmoore) [inline docs/contributing.rst:188]: :100:

**Comment by joshmoore** [comment]
Deactivating this button:
Screen Shot 2022-10-15 at 19 13 21
wasn't enough. I'm adding a .readthedocs.yaml file.

(On the plus side, pre-commit.ci is activated! :+1:)

**Comment by joshmoore** [comment]
Screen Shot 2022-10-15 at 19 24 01
:tada:

**Comment by joshmoore** [comment]
8 Likes and no objections -- https://twitter.com/zarr_dev/status/1579838254379339776

Let's do it!

**Comment by DimitriPapadopoulos** [comment]
The LGTM.com analysis can be removed now that GitHub code scanning has been added:
[The next step for LGTM.com: GitHub code scanning!](https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/)

**Comment by DimitriPapadopoulos** [comment]
Why remove section `[codespell] ` from `setup.cfg`?

Is it OK to put it back in `pyproject.toml` now that codespell supports it? See https://github.com/codespell-project/codespell/pull/2523, included in latest codespell release 2.2.2. It's much better to maintain the list of false positives in `pyproject.toml`, so that it is used when running codespell from the command line.

**Comment by Saransh-cpp** [comment]
Zarr now uses codespell as a pre-commit hook 😉 which means that the spell check is automated and should be performed using `pre-commit`.

See -
https://github.com/zarr-developers/zarr-python/blob/4362bb438f537b37ebd271506b774dfa31f5e68e/.pre-commit-config.yaml#L17-L21

**Comment by DimitriPapadopoulos** [comment]
I see. You could maintain codespell options in `pyproject.toml` and run it as a pre-commit hook nevertheless.

At some point, I expect a false positive to be introduced in new code. I feel that being able to run codespell directly, with the same options, will help understand and fix the problem.

**Comment by joshmoore** [comment]
Seems if the options can be in pyproject.toml rather than .pre-commit-config.yaml that's optimal.

> The LGTM.com analysis can be removed now that GitHub code scanning has been added:

@DimitriPapadopoulos: just to be clear, I'll remove the app, yeah?

**Comment by DimitriPapadopoulos** [comment]
> Seems if the options can be in pyproject.toml rather than .pre-commit-config.yaml that's optimal.

Yes, but I need to fix [a few issues I have with codespell](https://github.com/codespell-project/codespell/issues/2532) first, to be able to test with the latest version 2.2.2, which supports `pyrpoject.toml`.

**Comment by Saransh-cpp** [comment]
IMO it would be better to keep the configuration of all the `pre-commit` hooks at one place - inside `.pre-commit-config.yaml`. You can always do -
```
pre-commit run codespell --all-files
```
to run codespell on all files.

**Comment by DimitriPapadopoulos** [comment]
The point is that codespell is not necessarily run from pre-commit.

**Comment by joshmoore** [comment]
@Saransh-cpp: I'd guess I'd counter with how much can be inside the pyproject.toml? :) @DimitriPapadopoulos has really been leading the lint-ification and so if we can keep the "running without pre-commit" use case, that'd be great.

**Comment by Saransh-cpp** [comment]
Yes, running it without pre-commit sounds good too! Any path that would make the process smoother!

**Comment by jakirkham** [comment]
Would be good to make the same changes to [Numcodecs]( https://github.com/zarr-developers/numcodecs ) 🙂

**Comment by DimitriPapadopoulos** [comment]
Also, I do not have much experience with `pre-commit`, but it doesn't work for me on Ubuntu 22.04:
```
$ pre-commit run codespell --all-files
[INFO] Installing environment for https://github.com/codespell-project/codespell.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/usr/bin/python3', '-mvirtualenv', '/home/username/.cache/pre-commit/repolbk82i8v/py_env-python3.9', '-p', 'python3.9')
return code: 1
expected return code: 0
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.9'
    
stderr: (none)
Check the log at /home/username/.cache/pre-commit/pre-commit.log
$ 
```
I am probably doing something stupid here.

**Comment by mzjp2** [comment]
Do you have Python 3.9 installed?

**Comment by DimitriPapadopoulos** [comment]
No, I work with Python 3.10.

**Comment by mzjp2** [comment]
That'll be why it's failing - we should probably loosen the `pre-commit` language version to just `python3`.

**Comment by mzjp2** [comment]
Yep, just change it to `python3` instead.

**Comment by jakirkham** [comment]
This was changed to `python3` along with a few other fixes that were need in PR ( https://github.com/zarr-developers/zarr-python/pull/1197 )

**Comment by jakirkham** [comment]
> The LGTM.com analysis can be removed now that GitHub code scanning has been added:
> [The next step for LGTM.com: GitHub code scanning!](https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/)

Done in PR ( https://github.com/zarr-developers/zarr-python/pull/1191 )

**Comment by joshmoore** [comment]
seealso: https://github.com/zarr-developers/zarr-python/issues/1209

**Pull request merged**